tests/integration/hsm-integration-test.cpp

```diff
@@ -0,0 +1,452 @@
+                openssl::EllipticCurvePointConversionForm::uncompressed);
+    } catch (const openssl::OpenSSLException &e) {
+        /* low level OpenSSL failure */
+        std::cerr << "Failure transforming EC key." << std::endl;
+        std::cerr << e.what() << std::endl;
+        exit(EXIT_FAILURE);
+    } catch (const MoCOCrWException &e) {
+        /* Possible reasons:
+         * - Key object doesn't contain an EC key
+         */
+        std::cerr << "Failure transforming EC key." << std::endl;
+        std::cerr << e.what() << std::endl;
+        exit(EXIT_FAILURE);
+    }
+    return EciesEncryptResult{
+            /* The serialization of the ephemeral key's public component is up to the implementer.
+             * The standard is not defining a format. Available formats in MoCOCrW are:
+             * - uncompressed (used here)
+             * - compressed
+             * - hybrid
+             */
+            ephemeralKey,
+            encryptedData,
+            encCtx->getMAC(),
+    };
+}
+
+void decryptData(const EciesEncryptResult &eciesData, const AsymmetricPrivateKey &privKey)
+{
+    /* We need
+     * - mac value
+     * - ephemeral key
+     * - ciphertext
+     * - private key
+     */
+
+    /* The standard for ECIES (IEEE 1363a-2004) doesn't specify the details of the different cipher
+     * blocks. Thus it is up to the user to define its own set of cipher blocks (hash-function, key
+     * derivation function, message authentication code, symmetric cipher) used for encryption and
+     * decryption. In the current example the following is used:
+     * - SHA512 everywhere where a hash function is required
+     * - X963(SHA512) as key derivation function
+     * - HMAC(SHA512) as message authentication code
+     * - AES CBC with PKCS padding, zero IV and 256 bit key size
+     * - Additionally salts for KDF and MAC can be set (optional value, default: empty)
+     *
+     * The defaults if the builder is invoked without any optional parameters are:
+     * - X963(SHA512) as key derivation function
+     * - HMAC(SHA512)
+     * - AES CBC with PKCS padding, zero IV and 256 bit key size
+     * - empty salts for KDF and MAC
+     */
+
+    std::shared_ptr<AsymmetricKey::Spec> spec = privKey.getKeySpec();
+    auto eccSpec = std::dynamic_pointer_cast<ECCSpec>(spec);
+    if (!eccSpec) {
+        std::cerr << "Given key is no ECC key." << std::endl;
+        exit(EXIT_FAILURE);
+    }
+
+    /* The underlying openssl function recognizes the used serialization format of the ephemeral key
+     * reading the first byte of the serialized data, which defines the format.
+     * The variable eccSpec contains the used elliptic curve of the ephemeral key */
+    auto _ephemeralKey = AsymmetricPublicKey::fromECPoint(eccSpec, eciesData.ephemeralKey);
+
+    /* Get the decryption context */
+    auto decCtx =
+            ECIESCtxBuilder{}  // This is optional. The default is X963 with SHA512
```

Comment:
Maybe we should not specify all the details in the context of this test here and just use the defaults where possible. Here, the focus is on the Mococrw <> HSM interplay and not at showcasing the configurability of ECIES.

Reply:
Done